### PR TITLE
Panama Withholding

### PR DESCRIPTION
**Pull request by hbto**
## Video Example:
https://drive.google.com/file/d/0B1w9gokcJL9hWHlNRDMzajFyN2M/view

## Todo List:
- [x] Mark Partners as Withholding Agent.
- [x] Field on Partner, Withholding Subject: To Discuss @nhomar, for reporting purpose it needs to be an integer, for business process needs to be a value, do we use a new record for this or do we wire the value to the integer value in Business Process.
- [x] Add wh_move_id on Invoice
- [x] Add server action to `open` node on Invoice Workflow to execute ITBMS Withholding
- [x] Withholding is only to Apply on Customer Invoice and Customer Refund - 9318d69
- [x] Fix Debit - Credit direction on Journal Items. - 9318d69
- [x] Reusing method compute_invoice_totals() on invoice. - 6f82f78
- [x] Omit Invoices from withholding, for example for petty cash. - 54638d8
- [x] When Withholding does not apply change the value from zero (current value) to 'na' because zero is not taken as an option to fill the field. - fc518da
- [x] Retrieve Accounting Configuration from Company and validate. - 19e09a8
- [x] Compute Withholding amount on each tax line base on Withholding Subject value on invoice 2ef3b5c
- [x] Withholding Agent Entitlement on invoice -    4245979
- [x] Marcar al Partner con campo para Agente de retención.
- [x] Propagate Withholding Subject from Sale Order to Invoice, when invoicing from Sales Order - c21d2d4
- [x] Propagate Withholding Subject from Sale Order to Invoice, when invoicing from Picking - d3cc67b
- [x] Taking away dependency on l10n_pa_reports - c866be3
- [x] Switching dependency toward l10n_pa_withholding from l10n_pa_reports - f9d8ba3
- [x] Add sale_stock to dependencies - 9934518
- [x] On Onchange Partner in Sales Order bring the Fiscal Information from AccountingPartner. - 8d64c37
- [x] Add Demo Data to be used in UnitTesting - 7f0180a1 - 20005fc - 33c81a9
- [x] UnitTesting Module - 7f0180a - 069d0d6 - 9fa03dc - 4987324 - 33c81a9 - f68d7a1 - 531f60f
- [x] Abstracting all duplicated fields in a common class/model - 7ad5250
- [x] Fiscal Info should only be available from Accounting Partner - 1419f48
- [x] Propagate field values to refund (Fiscal Info) when creating refund - de96186
- [x] Reconciliation Process, stitching together move_id on Invoice with wh_move_id from withholding.
- Without previous reconciliations on Invoice. [DONE]
- With previous reconciliatioin on Invoice. [DONE]
- [x] Adding Journal at company Settings to use when Withholding Taxes - 68c9a4f
- [x] Module Translation into Spanish -     4e6aaf0
- [x] There shall be some way to mark taxes as withholdable - ce45b8e
- [x] Process for cancellation of invoice - 80e9752
- [x] Keep track of wh_move_id.name (store it in invoice) to reuse it after cancellation - 6753199
- [x] Withholding is be created when using Withholding Subject in ('5', '6') - bae60c4 - da63cfd
- [ ] Field in account.move.line for afterwards Withholding Report purposes  -

## Other Pending Items
- [ ] Review Withholding Case Scenarios - Revisar Casos de retenciones.

### CODE REVIEW

File: l10n_pa_withholding/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+{
+    "name": "Panama Withholding",
+    "version": "8.0.0.1.0",
+    "author": "Vauxoo",
+    "category": "Generic Modules",
+    "website": "http://www.vauxoo.com/",
+    "license": "AGPL-3",
+    "depends": [
+            "l10n_pa_reports",
```
Comment (nhomar) [inline l10n_pa_withholding/__openerp__.py]: Why depend of reports? I was expecting this as a procedural modules only, in all case the one that should depend of this one is the l10n_pa_report

File: l10n_pa_withholding/model/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    @api.model
+    def wh_tax_account(self, line):
+        """Retrieves the account to use as Withholding Account"""
+        return line.account_id.id
+
+    @api.model
+    def wh_move_line_get_item(self, line):
+        sign = 1
+        if line.invoice_id.type in ('out_invoice', 'in_invoice'):
```
Comment (nhomar) [inline l10n_pa_withholding/model/account_invoice.py]: this methodology is only valid for out_invoice and out_refund not for in_invoice that case has not been discussed yet.

File: l10n_pa_withholding/model/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            'name': line.name.split('\n')[0][:64],
+            'account_id': self.wh_tax_account(line),
+            'price': sign * line.amount,  # TODO: Insert Withheld value here!
+            'tax_code_id': line.tax_code_id.id,
+            'tax_amount': line.amount,
+        }
+
+    @api.model
+    def wh_move_line_get(self):
+        """Creates Withholding for taxes in invoice"""
```
Comment (nhomar) [inline l10n_pa_withholding/model/account_invoice.py]: THe agreement was create an extra account move, not an extra tax.line, why are you coding an extra tax.line?

**Comment by nhomar** [comment]
I can't see:
- Flag in partner to manage the withholding process.
- Process is not what we discussed, we discussed an extra account move not an extra tag.

Please fix.

File: l10n_pa_withholding/__openerp__.py (hunk 1 above)
Comment (hbto) [reply l10n_pa_withholding/__openerp__.py]: Preliminar proposal not set in stone

File: l10n_pa_withholding/model/account_invoice.py (hunk 2 above)
Comment (hbto) [reply l10n_pa_withholding/model/account_invoice.py]: then just need to override the offending code

File: l10n_pa_withholding/model/account_invoice.py (hunk 3 above)
Comment (hbto) [reply l10n_pa_withholding/model/account_invoice.py]: As I have mentioned before to you privately, _there are two commits_ both are to be reviewed widen your sight

**Comment by hbto** [comment]
- [ ] Field on Partner, Withholding Subject: To Discuss @nhomar, for reporting purpose it needs to be an integer, for business process needs to be a float, that ranges from 1 to 100, do we use a new record for this or do we wire the value to the integer value in Business Process.

**Comment by nhomar** [comment]
On Tue, Oct 25, 2016 at 3:47 PM, Hbto [ Vauxoo ] notifications@github.com
wrote:

> - Field on Partner, Withholding Subject: To Discuss @nhomar
>   https://github.com/nhomar, for reporting purpose it needs to be an
>   integer, for business process needs to be a value, do we use a new record
>   for this or do we wire the value to the integer value in Business Process.
> 
> It should be a selection with possible Values:

No Withhold
Percentage 1
Percentage 2
..
...
....
Percentage N

## 

Nhomar Hernandez
CEO Vauxoo.
Site: http://vauxoo.com
Twitter: @nhomar
Blog: http://nhomar.com
Github User: https://github.com/nhomar
Odoo Gold Partner
Skype: nhomar00 (Envia mail previo no lo superviso siempre).
HangOut: nhomar@vauxoo.com
Móvil Venezuela:
+58 4144110269
Móvil México:
+52 1 4773933942

**Comment by hbto** [comment]
@nhomar 
Do we need to propagate info from SaleOrder to Invoice through path
SaleOrder > Picking > Invoice?

**Comment by hbto** [comment]
@nhomar 

> Do we need to propagate info from SaleOrder to Invoice through path
> SaleOrder > Picking > Invoice?

Already done here:

Propagate Withholding Subject from Sale Order to Invoice, when invoicing from Picking - d3cc67b

File: l10n_pa_withholding/__openerp__.py [hunk 4]
```diff
@@ -10,6 +10,7 @@
     "depends": [
             "l10n_pa_reports",
             "sale",
+            "stock",
```
Comment (nhomar) [inline l10n_pa_withholding/__openerp__.py]: can you add sale_stock to dependencies pliz

Comment (hbto) [reply l10n_pa_withholding/__openerp__.py]: @nhomar 
ok, got it

Comment (hbto) [reply l10n_pa_withholding/__openerp__.py]: @nhomar Done at 9934518

File: l10n_pa_withholding/__openerp__.py (hunk 1 above)
Comment (hbto) [reply l10n_pa_withholding/__openerp__.py]: @nhomar 
Already fixed @ c866be3

**Comment by nhomar** [comment]
Según el plan:
- Campo en Aml de objeto de retención para efectos de reportes posteriores?

Puede ser un diario no? o un tipo de diario?

**Comment by hbto** [comment]
@nhomar 

Following Items are pending:
- [x] Module Translation into Spanish (DOING)
- [x] Field in account.move.line for afterwards Withholding Report purposes  - Campo en Aml de objeto de retención para efectos de reportes posteriores?
  Can we meet to further discuss this statement of yours
  
  > Puede ser un diario no? o un tipo de diario?
  > I have some doubts
- [x] Will we use the same journal than that of the invoice + a String for Journal Entry Naming or a new Journal?
  Do you have any idea or can we cross this item, getting rid of it.
- [ ] How to proceed on cancellation of invoice?
  What would you suggest?

**Comment by hbto** [comment]
@nhomar Could you review and provide further guidance.

Regards

File: l10n_pa_withholding/model/account_invoice.py [hunk 5]
```diff
@@ -191,3 +201,90 @@ def wihholding_reconciliation(self):
             line_ids.reconcile_partial()
 
         return True
+
+
+class AccountInvoiceTaxWh(models.Model):
+    """Invoice Tax Withholding"""
+    _name = 'account.invoice.tax.wh'
+    _inherit = 'account.invoice.tax'
```
Comment (nhomar) [inline l10n_pa_withholding/model/account_invoice.py:243]: don't forget _rec_name

**Review by nhomar — CHANGES_REQUESTED** [review]

File: l10n_pa_withholding/model/account_invoice.py (hunk 5 above)
Comment (hbto) [reply l10n_pa_withholding/model/account_invoice.py:243]: There's no need for `_rec_name` I am prototyping from `account.invoice.tax` which already has `name`.

Or should I?
models

**Comment by hbto** [comment]
@nhomar 
https://drive.google.com/file/d/0B1w9gokcJL9hWHlNRDMzajFyN2M/view

**Comment by hbto** [comment]
Method to compute taxes needs to be fixed,
When wh_subject is `5` or `6` the withholding is performed based on
the whole invoice amount_total, rather than tax based.

**Pull request merged**